libminifi/include/utils/MinifiConcurrentQueue.h

```diff
@@ -119,10 +133,22 @@ class ConditionConcurrentQueue : private ConcurrentQueue<T> {
     }
   }
   
-  bool dequeue(T& out) {
+  bool dequeueWait(T& out) {
+    std::unique_lock<std::mutex> lck(this->mtx_);
+    cv_.wait(lck, [this, &lck]{ return !running_ || !this->emptyImpl(lck); });  // Only wake up if there is something to return or stopped 
+    return running_ && ConcurrentQueue<T>::tryDequeueImpl(lck, out);
+  }
+
+  template< class Rep, class Period >
+  bool dequeueWaitFor(T& out, const std::chrono::duration<Rep, Period>& time) {
```

Comment:
I like that you used a template here, not just a concrete duration type like `milliseconds`, which is more common in the code base.